### ISSUE TITLE
Weird behavior for push notifications

### ISSUE THREAD

**Issue by MaxBleggi** [ParsePlatform/Parse-SDK-Android#11]
I read through the getting started tutorial and then the trouble shooting section to no avail. 

Before I show you specifics, let me explain the issue.
Yesterday I tried to implement your push notification service but I could not receive any. So I kept trying and none were received. I tried scheduling and making them now but it didn't work. So I went to bed to try again tomorrow. 

At 6 in the morning I was awoken to my phone buzzing. I looked at it and there were all the dozen notifications I had tried to send. Why did I receive them like 8 hours later? So later in the day I try to do it again and no luck. Until, sure enough, a few hours later when I was trying something else I received like a dozen push notifications (all of them the ones I had tried to send earlier) So right after this happened I tried sending another notification and  bam. Right after I pushed the send button, I received it. Awesome! So I thought the problem had just fixed itself until I tried sending another push a little while ago and nothing happened. I've tried to send like 20 different pushes on different channels and nothing. (It should be noted that the night I first tried it I was on the default Broadcast channel and today I tried switching between that and a custom channel).

I have no idea what's wrong. I'm assuming it's either on my end or a possible bug in the SDK. So here is my code.

Here's my starting class
```java
public class InitializeParse extends Application
{
    public void onCreate()
    {
        super.onCreate();
        Parse.enableLocalDatastore(getApplicationContext());
        Parse.initialize(this, "PRIVATE_INFO", "PRIVATE_INFO");
        ParseInstallation.getCurrentInstallation().saveInBackground();

        ParsePush.subscribeInBackground("", new SaveCallback()
        {
            @Override
            public void done(ParseException e)
            {
                if (e == null)
                {
                    Toast.makeText(InitializeParse.this, "success!", Toast.LENGTH_LONG).show();
                } else
                {
                    Toast.makeText(InitializeParse.this, "not working", Toast.LENGTH_LONG).show();
                }
            }
        });
    }
}
```
It always toasts "success!"

Here's my manifest

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest
    xmlns:android="http://schemas.android.com/apk/res/android"
    package="maxbleggi.afstudentplanner" >

    <uses-permission android:name="android.permission.GET_ACCOUNTS" />
    <uses-permission android:name="android.permission.READ_PROFILE" />
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.WAKE_LOCK" />
    <uses-permission android:name="android.permission.VIBRATE" />
    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED" />
    <uses-permission android:name="com.google.android.c2dm.permission.RECEIVE" />

    <permission android:protectionLevel="signature"
        android:name="maxbleggi.afstudentplanner.permission.C2D_MESSAGE"/>
    <uses-permission android:name="maxbleggi.afstudentplanner.permission.C2D_MESSAGE" />

    <application
        android:name=".utils.InitializeParse"
        android:allowBackup="true"
        android:icon="@mipmap/ic_launcher"
        android:label="@string/app_name"
        android:theme="@style/AppTheme" >
        <activity
            android:name=".activities.MainActivity"
            android:label="@string/app_name" >
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />

                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
        <activity
            android:name=".activities.RegisterActivity"
            android:label="@string/title_activity_register" >
        </activity>
        <activity
            android:name=".activities.RegisterPart_2Activity"
            android:label="@string/title_activity_register_part_2" >
        </activity>
        <activity
            android:name=".activities.RegisterPart2TeacherActivity"
            android:label="@string/title_activity_register_part2_teacher" >
        </activity>
        <activity
            android:name=".activities.EventListActivity"
            android:label="@string/title_activity_event_list"
            android:parentActivityName=".activities.MainActivity" >
            <meta-data
                android:name="android.support.PARENT_ACTIVITY"
                android:value=".activities.MainActivity" />
        </activity>
        <activity
            android:name=".activities.NoteListActivity"
            android:label="@string/title_activity_note_list"
            android:parentActivityName=".activities.EventListActivity" >
            <meta-data
                android:name="android.support.PARENT_ACTIVITY"
                android:value=".activities.EventListActivity" />
        </activity>
        <activity
            android:name=".activities.SplashScreenActivity"
            android:label="@string/title_activity_splash_screen" >
        </activity>
        <activity
            android:name=".activities.RegisterPart_3TeacherActivity"
            android:label="@string/title_activity_register_part_3_teacher" >
        </activity>

        <!--<meta-data android:name="com.parse.push.notification_icon"/>
                   android:resource="@drawable/ic_stat_bt"-->

        <service android:name="com.parse.PushService" />
        <receiver android:name="com.parse.ParseBroadcastReceiver">
            <intent-filter>
                <action android:name="android.intent.action.BOOT_COMPLETED" />
                <action android:name="android.intent.action.RECEIVE_BOOT_COMPLETED" />
                <action android:name="android.intent.action.USER_PRESENT" />
            </intent-filter>
        </receiver>
        <receiver android:name="com.parse.ParsePushBroadcastReceiver" android:exported="false">
            <intent-filter>
                <action android:name="com.parse.push.intent.RECEIVE" />
                <action android:name="com.parse.push.intent.DELETE" />
                <action android:name="com.parse.push.intent.OPEN" />
            </intent-filter>
        </receiver>
        <receiver android:name="com.parse.GcmBroadcastReceiver" android:permission="com.google.android.c2dm.permission.SEND">
            <intent-filter>
                <action android:name="com.google.android.c2dm.intent.RECEIVE" />
                <action android:name="com.google.android.c2dm.intent.REGISTRATION" />
                <category android:name="maxbleggi.afstudentplanner" />
            </intent-filter>
        </receiver>

    </application>

</manifest>
```

The parse notification dashboard always tells me they were sent successfully and there was 1 push sent (that is the current number of installations I have (me)) 

According to the parse tutorial and docs, everything should be going fine. I don't know what is causing this

**Comment by ijazfahad** [comment]
This happened to me too. Try closing the app fully before sending the push. That worked for me and I got the push on time, but I know this isn't the real solution.

**Comment by MaxBleggi** [comment]
I closed, reopened, reinstalled, deleted all users, everything. Nothing seems to work.

I hope someone from the parse team can clear the confusion

**Comment by ijazfahad** [comment]
are you using an emulator or a real phone? that could be the problem maybe? i don't know if emulators get push notification or not.

**Comment by stanleyw** [comment]
1. Could you try this on the latest SDK and see whether it repros?

2. If so, this could also be a backend issue.  We had some push system delays on Aug 11th (https://status.parse.com/).  We continue to use the FB bug tool to track backend issues, where you can confidentially submit information about the exact app that you are having issues on.  Could you file an issue there?  https://www.parse.com/help

More about SDK bug/repro guidelines at: https://github.com/ParsePlatform/Parse-SDK-Android/blob/master/CONTRIBUTING.md#bugs   We plan to use GitHub issues for SDK-specific bugs only.

The Android simulator should be able to receive pushes, but it's sometimes more flaky than a real device.  I recommend testing on a real device whenever possible.

**Comment by grantland** [comment]
This actually sounds most likely a backend or GCM issue as we don't actually persist anything on the device for push notifications, so you'd either get it right away or not at all. In addition to what Stanley mentioned, you can debug with the following:

* Enable logging with `Parse.setLogLevel(int)` and observe whether GCM or PPNS is being used in your error case.
* Implement your own `BroadcastReceiver` with an intent filter for `com.google.android.c2dm.intent.RECEIVE` (and remove that filter for `GcmBroadcastReceiver`) with some logging in `onReceive()`. This will get you as raw of a connection to the push pipeline as possible and you can see whether or not the device is actually receiving push notifications.
* View device logcats. There might be some info from Google Play Services on when it it actually notifying your application of push notifications.

If you can see the application being notified immediately, it might be an issue in the SDK. Otherwise, if the Google Play Services isn't notifying us, it's most likely a delay in our backend or GCM.

**Comment by MaxBleggi** [comment]
Before I continue, I want to add that again at 3 in the morning I received about 30 pushes all at once.

- yes I am using a real device

@stanleyw 

1.  today I switched from the parse*.jar file to the  ``` 'com.parse:parse-android:1.10.0' ``` library. I put that dependency in my build.gralde, deleted the old dependencies for parse, deleted the .jar in my libs folder, and for extra measure I cleaned the project. I still had the same issues.

2. After some further debugging with your guys advise, it looks like it is indeed not an SDK issue after all. I'll file an issue there.

@grantland 

- put ``` Parse.setLogLevel(Parse.LOG_LEVEL_DEBUG)``` as well as making a simple custom receiver. I'll show the code below. Either I did it wrong or I'm not receiving any pushes. (I'm pretty sure it's the latter) I did not see any indication of receiving a push. So it looks like it's not an SDK issue after all.

Here's my manifest
```xml
        <receiver android:name="utils.PushReceiver" >
            <intent-filter>
                <action android:name="com.google.android.c2dm.intent.RECEIVE" />
            </intent-filter>
        </receiver>
```

my custom receiver
```java
public class PushReceiver extends BroadcastReceiver
{
    @Override
    public void onReceive(Context context, Intent intent)
    {
        Log.d("custom", "PUSH RECEIVED!!!");
    }
}
```
 (I've never made a receiver before but I'm fairly certain that it should at least hit the log line if it were to receive a push)


Thanks for the help and quick responses. It looks like I'll be filing the issue to a different source

**Comment by grantland** [comment]
Yeah unfortunately it looks like a backend or GCM issue :(

**Comment by egenvall** [comment]
Did you find any solution to the problem? I have the same issue with 1.10.
Just to see if I did something wrong I tried the push example
https://github.com/ParsePlatform/PushTutorial
with 1.4.3 that uses the depecated method
PushService.setDefaultPushCallback(this, MainActivity.class);

and it is receiving all my Pushes from the website instantly, without delay. However I have yet to succeed using the newest SDK. Kind of frustrating

**Comment by MaxBleggi** [comment]
I went to facebook to find out if it was a backend issue. here's a link to my debug report: https://developers.facebook.com/bugs/1469047606730343

It was determined that it was something on my device not connecting to GCM or something. 

This is a quote from the report after some attempted debuging

"Our engineers had a look at this issue and they had the following comments:

"Pushes Sent" in push console means we successfully handed the push notification to vendors that are beyond in our control (APNS for iOS and GCM for android). We don't know when the message is actually delivered and could do nothing about it.

Regarding the delay seen by the developer, I think the issue is most likely on his device and/or the connection between his device and GCM"


They sent the push successfully onto GCM (for android). After that they have no control over it. And as seen here. It is not an SDK issue. So it is something between GCM and my device.

After tinkering with an emulator to get it to accept pushes, I tried to send some and they all sent just fine. It appeared it was my device that was the problem
